src/java.base/share/classes/sun/security/rsa/RSAPrivateCrtKeyImpl.java

```diff
@@ -276,16 +271,32 @@ protected long getNativePtr() {
         byte[] dQ_2c   = dQ.toByteArray();
         byte[] qInv_2c = qInv.toByteArray();
 
-        nativeRSAKey = nativeCrypto.createRSAPrivateCrtKey(n_2c,n_2c.length, d_2c, d_2c.length, e_2c, e_2c.length,
+        newNativePtr = nativeCrypto.createRSAPrivateCrtKey(n_2c,n_2c.length, d_2c, d_2c.length, e_2c, e_2c.length,
                 p_2c, p_2c.length, q_2c, q_2c.length,
                 dP_2c, dP_2c.length, dQ_2c, dQ_2c.length, qInv_2c, qInv_2c.length);
-        return nativeRSAKey;
+        
+        return newNativePtr;
+    }
+
+    protected long getNativePtr() {
+        Long ptr = keyQ.poll();
+        if (ptr == null) {
+            return RSAPrivateKey_generate();
+        }
+        return ptr;
+    }
+
+    protected void returnNativePtr(Long ptr) {
```

Comment:
The way this is used, `ptr` will never be `null`. I suggest you change the parameter type to `long` (to match the caller) and do the boxing here. Same comment applies to similar code in `RSAPublicKeyImpl`.

Reply:
Hello Sir, sorry i am not sure if i understand boxing correctly. is this what you suggest for the modification? 
`protected void returnNativePtr(long ptr) {`
      `      keyQ.add(ptr);`
`    }`

Reply:
Yes, that's what I meant. There's an implicit use of `Long.valueOf()` in `keyQ.add(ptr);` so it's as if you wrote `keyQ.add(Long.valueOf(ptr));`.